File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/value/AttValue.java
package ru.citeck.ecos.records3.record.op.atts.service.value;

import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records3.record.op.atts.service.value.impl.SimpleAttEdge;

import java.util.Collections;

/**
 * Attribute value. Used to get attributes by schema.
 *
 * @author Pavel Simonov
 */
public interface AttValue {

    /**
     * Value identifier. Can be RecordRef or String
     */
    default Object getId() {
        return null;
    }

    default String getDispName() throws Exception {
        return asText();
    }

    // ===== get as =====

    default Object as(@NotNull String type) throws Exception {
        return null;
    }

    /**
     * String representation.
     */
    default String asText() throws Exception {
        return toString();
    }

    default Double asDouble() throws Exception {
        String str = asText();
        return str != null ? Double.parseDouble(str) : null;
    }

    default Boolean asBool() throws Exception {
        String txt = asText();
        if (Boolean.TRUE.toString().equals(txt)) {
            return true;
        } else if (Boolean.FALSE.toString().equals(txt)) {
            return false;
        }
        return null;
    }

    default Object asJson() throws Exception {
        return asText();
    }

    // ===== /get as =====

    default RecordRef getTypeRef() throws Exception {
        return RecordRef.EMPTY;
    }

    default boolean has(@NotNull String name) throws Exception {
        return false;
    }

    /**
     * Get value attribute.
     */
    default Object getAtt(@NotNull String name) throws Exception {
        return Collections.emptyList();
    }

    default AttEdge getEdge(@NotNull String name) throws Exception {
        return new SimpleAttEdge(name, this);
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/value/AttValue.kt
package ru.citeck.ecos.records3.record.op.atts.service.value

import ru.citeck.ecos.commons.json.Json
import ru.citeck.ecos.records2.RecordRef
import ru.citeck.ecos.records3.record.op.atts.service.value.impl.SimpleAttEdge

/**
 * Attribute value. Used to get attributes by schema.
 *
 * @author Pavel Simonov
 */
interface AttValue {

    fun getId(): Any? = null

    @Throws(Exception::class)
    fun getDisplayName(): String? = asText()

    @Throws(Exception::class)
    fun asText() : String? = toString()

    @Throws(Exception::class)
    fun getAs(type: String) : Any? = null

    @Throws(Exception::class)
    fun asDouble(): Double? = asText()?.toDouble()

    @Throws(Exception::class)
    fun asBoolean(): Boolean? {
        return when (asText()) {
            "true" -> true
            "false" -> false
            else -> null
        }
    }

    @Throws(Exception::class)
    fun asJson(): Any? = Json.mapper.read(asText())

    @Throws(Exception::class)
    fun has(name: String): Boolean = false

    @Throws(Exception::class)
    fun getAtt(name: String): Any? = null

    @Throws(Exception::class)
    fun getEdge(name: String): AttEdge? = SimpleAttEdge(name, this)

    @Throws(Exception::class)
    fun getType() : RecordRef = RecordRef.EMPTY
}
